src/shared/cgroup-setup.c

```diff
@@ -18,6 +18,32 @@
 #include "user-util.h"
 #include "fd-util.h"
 
+static bool cg_is_unified_possible(void) {
+        _cleanup_set_free_free_ Set *controllers = NULL;
+        int r;
+
+        /* If we are running in a container, we need to check if the host
+         * is using cgroup1 (or hybrid), in which case it's not possible
+         * for us to use cgroup2. We don't actually check if we're running
+         * in a container here, as detect_container() might rely on cgroups
+         * being mounted, and we may be called before that happens.
+         *
+         * An alternate way to check this might be to read /proc/self/cgroup
+         * to see if there is more than 1 (index 0) entry, since if cgroup1
+         * is in use our process cgroup file will list all the controllers
+         * in use by cgroup1, in addition to the default/root controller.
+         *
+         * Here, we just check /proc/cgroups for any controllers that are
+         * in use by cgroup1 currently. */
+        r = cg_kernel_controllers_all(&controllers, true, true);
```

Comment:
that's a bit too magic for case. what happens if you run a cgroup 1 host with all controllers disabled? 

Really, the only safe way for this to work is if the hierarchy is pre-mounted correctly by the container manager, or we are told via kernel cmdline param. But guesswork is guesswork.

Most container managers should get this right today, and not require any work. Which container manager do you have in mind here which doesn't get this right? can#t you fix that to make things just work?

My guess is that if the container manager doesn't get this right it's cgroups logic is pretty messed up in general.